src/transformers/utils/model_parallel_utils.py

```diff
@@ -12,10 +12,186 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+import copy
+import importlib
+from collections import defaultdict
 from math import ceil
 
 
+class TPInfo(object):
+    def __init__(
+        self,
+        *name,
+        combined_qkv: bool = False,
+        reverse: bool = False,
+    ):
+        self.name = name
+        self.combined_qkv = combined_qkv
+        self.reverse = reverse
+
+    def __str__(self):
+        return f"{self.__class__.__qualname__}({self.name})"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+Col = type("COLUMN", (TPInfo,), {"code": "Col"})
+Row = type("ROW", (TPInfo,), {"code": "Row"})
+Update = type("UPDATE", (TPInfo,), {"code": "Update"})
+
+
+class TPMapping(object):
```

Comment:
And here `TensorParallelismMapping`. We also need a docstring to explain what this does and how to expand it.